### PR TITLE
Sorted and rearranged core whitelists via generation

### PR DESCRIPTION
**Pull request by lzap**
This PR rearanges whitelists in a reasonable way so it's easier to maintain. In several commits I break entries into multiple lines, sort them and rearrange them a bit with no changes performed. Then I add Ruby core and ActiveSupport core methods to the default_methods array. And finally, new rake task `generate` is added which generates whitelists and the last commit finally makes changes to the whitelists which need to be carefully reviewed.

The idea is to use generation approach for maintaining those whitelists, everytime there is a new Ruby release we can easily generate those. The rake task contains list of "dangerous" methods which I think should never go into Safemode therefore we do not need to bother with them during review.

I suggest to do review commit-by-commit as I was trying hard to make this as clear as possible. The last commit is the most important one - contains the script and the important diff - changes to the whitelists. I the review will go on, I will be amending only the last commit until we agree.

I will drop few comments now - these are still open questions.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -74,3 +72,107 @@ Rake::RDocTask.new do |rdoc|
...
+  require "safemode"
+  default_methods = Safemode.class_variable_get("@@default_methods")
+  dangerous_methods = %w(
+  !~
+  =~
+  class
+  class_eval
+  clone
+  deep_clone
+  deep_dup
```
Comment (lzap) [inline Rakefile]: Cloning/duping is probably not dangerous however it was never whitelisted except `duplicable?` which went away (does it make sense when cloning is not allowed?).

File: Rakefile [hunk 2]
```diff
@@ -74,3 +72,107 @@ Rake::RDocTask.new do |rdoc|
...
+  define_singleton_method
+  display
+  dup
+  enum_for
+  extend
+  gem
+  hash
```
Comment (lzap) [inline Rakefile]: Just to be on the safe side.

File: Rakefile [hunk 3]
```diff
@@ -74,3 +72,107 @@ Rake::RDocTask.new do |rdoc|
...
+  __id__
+  inspect
```
Comment (lzap) [inline Rakefile]: Inspecting is probably safe, however we do not want "attackers" to inspect things.

File: Rakefile [hunk 4]
```diff
@@ -74,3 +72,107 @@ Rake::RDocTask.new do |rdoc|
...
+  instance_eval
+  instance_exec
+  instance_of?
+  instance_variable_defined?
+  instance_variable_get
+  instance_variables
+  instance_variable_set
+  itself
```
Comment (lzap) [inline Rakefile]: Not sure about `itself` but want to be safe.

File: Rakefile [hunk 5]
```diff
@@ -74,3 +72,107 @@ Rake::RDocTask.new do |rdoc|
...
+  tainted?
+  tap
+  to_enum
+  trust
+  try
+  untaint
+  untrust
+  untrusted?
+  with_options
+  yield_self
```
Comment (lzap) [inline Rakefile]: Both `tap` and `yield_self` are probably safe, but I was not sure.

File: lib/safemode/core_jails.rb [hunk 6]
```diff
@@ -94,56 +94,114 @@ def core_jail_class_methods(klass)
...
+    find
+    find_all
+    find_index
     first
+    flat_map
     flatten
     flatten!
-    hash
+    grep
+    grep_v
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: Enables regexps essentially, I know that there is a TODO comment in the codebase like "are regexps safe"? Would be good idea to answer it here and also allow `=~`.

File: lib/safemode/core_jails.rb [hunk 7]
```diff
@@ -94,56 +94,114 @@ def core_jail_class_methods(klass)
...
     fetch
     fill
+    find
+    find_all
+    find_index
     first
+    flat_map
     flatten
     flatten!
-    hash
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: I see `hash` was actually whitelisted, I can probably keep it if you confirm.

File: lib/safemode/core_jails.rb [hunk 8]
```diff
@@ -94,56 +94,114 @@ def core_jail_class_methods(klass)
   # whitelisted methods for core classes ... kind of arbitrary selection
   @@methods_whitelist = {
   'Array' => %w(
+    all?
     any?
+    append
     assoc
     at
-    blank?
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: Some methods were moved to the `default_methods` so they are deleted from individual class lists.

File: lib/safemode/core_jails.rb [hunk 9]
```diff
@@ -94,56 +94,114 @@ def core_jail_class_methods(klass)
...
     inject
     insert
     join
+    keep_if
     last
+    lazy
     length
     map
     map!
-    nitems
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: Removed in 1.9, time to drop it?

File: lib/safemode/core_jails.rb [hunk 10]
```diff
@@ -153,137 +211,309 @@ def core_jail_class_methods(klass)
...
+    drop_while
     each
+    each_cons
+    each_entry
     each_key
     each_pair
+    each_slice
     each_value
+    each_with_index
+    each_with_object
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: These guys made me to do the patch...

File: lib/safemode/core_jails.rb [hunk 11]
```diff
@@ -153,137 +211,309 @@ def core_jail_class_methods(klass)
...
+    lazy
     length
+    map
+    max
+    max_by
     member?
     merge
     merge!
-    present?
-    rec_merge!
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: No idea what this is from, probably some custom gem.

File: lib/safemode/core_jails.rb [hunk 12]
```diff
@@ -153,137 +211,309 @@ def core_jail_class_methods(klass)
...
-    chop
-    chop!
     include?
     index
     insert
     intern
-    iseuc
-    issjis
-    isutf8
-    kconv
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: These are from http://ruby-doc.org/stdlib-2.3.6/libdoc/nkf/rdoc/String.html - need to figure this out, probably can whitelist them in some manual way.

File: lib/safemode/core_jails.rb [hunk 13]
```diff
@@ -153,137 +211,309 @@ def core_jail_class_methods(klass)
...
+    grapheme_clusters
+    grep
     gsub
     gsub!
-    hash
     hex
-    chomp
-    chomp!
-    chop
-    chop!
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: Oh this is just a problem with sort - I sorted using Czech locale and in our language "ch" is actually one characters, it's sorted differently. See above.

File: lib/safemode/core_jails.rb [hunk 14]
```diff
@@ -300,75 +530,98 @@ def core_jail_class_methods(klass)
...
     to_f
     to_i
-    tojis
-    tosjis
+    to_r
     to_str
     to_sym
-    toutf16
-    toutf8
-    to_xs
```
Comment (lzap) [inline lib/safemode/core_jails.rb]: Bunch of other methods from http://ruby-doc.org/stdlib-2.3.6/libdoc/nkf/rdoc/String.html and I will add them later.

File: lib/safemode/core_jails.rb [hunk 15]
```diff
@@ -239,12 +239,12 @@ def core_jail_class_methods(klass)
     abs
     blank?
     ceil
-    chr
     coerce
     div
     divmod
     downto
     floor
+    chr
```
Comment (ares) [inline lib/safemode/core_jails.rb]: which locale did you use for sorting? :-) feels like czech `ch`

File: lib/safemode/core_jails.rb [hunk 16]
```diff
@@ -65,19 +65,26 @@ def core_jail_class_methods(klass)
...
     methods
     new
     nil?
     not
+    presence
+    present?
     to_a
     to_jail
     to_param
+    to_param
```
Comment (ares) [inline lib/safemode/core_jails.rb]: seems as dup of line above

File: Rakefile [hunk 17]
```diff
@@ -74,3 +72,107 @@ Rake::RDocTask.new do |rdoc|
...
+  __send__
+  send
+  silently
+  singleton_class
+  singleton_method
+  singleton_method_added
+  singleton_methods
+  taint
+  tainted?
+  tap
```
Comment (ares) [inline Rakefile]: this is probably safe and could be useful

File: Rakefile (hunk 5 above)
Comment (ares) [reply Rakefile]: I'd allow

**Comment by ares** [comment]
few comments, looks reasonable

File: Rakefile (hunk 2 above)
Comment (lzap) [reply Rakefile]: Ok I am going to put `hash` back in the upcoming amend.

File: Rakefile (hunk 5 above)
Comment (lzap) [reply Rakefile]: Ok good idea.

File: lib/safemode/core_jails.rb (hunk 7 above)
Comment (lzap) [reply lib/safemode/core_jails.rb]: Keeping.

File: lib/safemode/core_jails.rb (hunk 12 above)
Comment (lzap) [reply lib/safemode/core_jails.rb]: Whitelisting them explicitly.

File: lib/safemode/core_jails.rb (hunk 15 above)
Comment (lzap) [reply lib/safemode/core_jails.rb]: Yes, I can confirm. Realized this too late, I am going to keep it in this Czech order and before squash I can fix this. Not big deal other than our `ch` character.

File: Rakefile (hunk 17 above)
Comment (lzap) [reply Rakefile]: I think so, putting it back.

**Comment by lzap** [comment]
Thanks, amended and pushed only the last comment.

**Comment by lzap** [comment]
Hello anyone? I'd like to look into other things but I have a bunch of PRs that nobody look at. @svenfuchs @kronn 

We use the gem in our project (https://theforeman.org) and we have a large and vibrant community, if you want to help with maintaining the project we can help. The core team is competent team with various skills and 2FA enabled everywhere!

**Comment by ares** [comment]
totally makes sense to me, @lzap please update and rebase the PR, happy to merge then so we can always generate fresh whitelists

**Pull request closed without merging**